worker/worker.go

```diff
@@ -850,18 +850,24 @@ func (w *worker) objectsHandlerGET(jc jape.Context) {
 			return
 		}
 
-		// randomize order of contracts so we don't always download from the same hosts
-		frand.Shuffle(len(contracts), func(i, j int) { contracts[i], contracts[j] = contracts[j], contracts[i] })
-
-		// move bad hosts to the back of the array, a bad host is a host that
-		// timed out, is out of funds or is gouging its prices
+		// make sure consecutive slabs are downloaded from hosts that performed
+		// well on previous slab downloads
 		sort.SliceStable(contracts, func(i, j int) bool {
-			return badHosts[contracts[i].HostKey] < badHosts[contracts[j].HostKey]
+			return performance[contracts[i].HostKey] < performance[contracts[j].HostKey]
 		})
 
-		badHostIndices, err := downloadSlab(ctx, w, cw, ss, contracts, w.downloadSectorTimeout)
-		for _, h := range badHostIndices {
-			badHosts[contracts[h].HostKey]++
+		timings, err := downloadSlab(ctx, w, cw, ss, contracts, w.downloadSectorTimeout)
+
+		// update historic host performance
+		//
+		// NOTE: we only update if we don't have a datapoint yet or if the
+		// returned timing differs from the default. This ensures we don't
+		// necessarily want to try downloading from all hosts and we don't reset
+		// a host's performance to the default timing.
+		for i, timing := range timings {
+			if _, exists := performance[contracts[i].HostKey]; !exists || timing != int64(defaultSectorDownloadTiming) {
+				performance[contracts[i].HostKey] = timing
```

Comment:
I think you may have read it wrong. I guess maybe the `NOTE` was unclear. 

We do overwrite datapoints, so we definitely do notice when hosts start performing better or worse.  
We overwrite if:
- there is currently no data point
- the data point differs from `defaultSectorDownloadTiming`

The `defaultSectorDownloadTiming` is a sentinel timing value that I use to prevent overwriting data points:

```
slab 1:
host1: 3s 
host2: 85ms
host3: 200ms -> `defaultSectorDownloadTiming` (host was not used)
host4: 200ms -> `defaultSectorDownloadTiming` (host was not used)
```

```
slab 2:
host1: 200ms -> `defaultSectorDownloadTiming` (host was not used)
host2: 3s        
host3: 320ms ->
host4: 200ms -> `defaultSectorDownloadTiming` (host was not used)
```

after slab1 host1 will have been pushed to the back, host2 is faster than the default so it will have been preferred when downloading slab2, if on slab2 it performs badly (like it does) we don't want to overwrite the host1 timing with the default value

I don't think it's overly complicated or over-engineered, it's just a sentinel value 🤷‍♂️ . I use it to prevent needlessly trying to download from all hosts available if we downloaded from fast hosts by coincidence (that's why the timings are all initialised to the default timing, otherwise we download from new hosts on consecutive slabs), and since I have it handy I can also use it to prevent overwriting historic performance for hosts that weren't used on consecutive download rounds.